Libraries/Opc.Ua.Client.ComplexTypes/Types/BaseComplexType.cs

```diff
@@ -368,268 +417,82 @@ int valueRank
         /// </summary>
         protected void EncodeProperty(
             IEncoder encoder,
-            PropertyInfo property,
-            int valueRank)
+            ComplexTypePropertyInfo property)
         {
-            EncodeProperty(encoder, property.Name, property, valueRank);
+            EncodeProperty(encoder, property.Name, property);
         }
 
         /// <summary>
         /// Encode a scalar property based on the property type.
         /// </summary>
-        private void EncodeProperty(IEncoder encoder, string name, PropertyInfo property)
+        private void EncodeProperty(IEncoder encoder, string name, PropertyInfo property, BuiltInType builtInType)
         {
             var propertyType = property.PropertyType;
-            if (propertyType == typeof(Boolean))
-            {
-                encoder.WriteBoolean(name, (Boolean)property.GetValue(this));
-            }
-            else if (propertyType == typeof(SByte))
-            {
-                encoder.WriteSByte(name, (SByte)property.GetValue(this));
-            }
-            else if (propertyType == typeof(Byte))
-            {
-                encoder.WriteByte(name, (Byte)property.GetValue(this));
-            }
-            else if (propertyType == typeof(Int16))
-            {
-                encoder.WriteInt16(name, (Int16)property.GetValue(this));
-            }
-            else if (propertyType == typeof(UInt16))
-            {
-                encoder.WriteUInt16(name, (UInt16)property.GetValue(this));
-            }
-            else if (propertyType.IsEnum)
+            if (propertyType.IsEnum)
```

Comment:
What would be the usecase that `propertyType.IsEnum` is true and  `builtInType` is not **Enumeration**

Reply:
added enum code coverage/tests and simplified. Enumeration can only be set if type is enum, ReadArray then returns the enum type instead of int32 type